benchmarks/src/jmh/java/io/grpc/benchmarks/CallOptionsBenchmark.java

```diff
@@ -0,0 +1,110 @@
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.benchmarks;
+
+import io.grpc.CallOptions;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Call options benchmark.
+ */
+@State(Scope.Benchmark)
+public class CallOptionsBenchmark {
+
+  @Param({"1", "2", "4", "8", "16"})
+  public int customOptionsCount;
+
+  private List<CallOptions.Key<String>> customOptions;
+
+  private CallOptions allOpts;
+  private List<CallOptions.Key<String>> shuffledCustomOptions;
+
+  /**
+   * Setup.
+   */
+  @Setup
+  public void setUp() throws Exception {
+    customOptions = new ArrayList<CallOptions.Key<String>>(customOptionsCount);
+    for (int i = 0; i < customOptionsCount; i++) {
+      customOptions.add(CallOptions.Key.of("name " + i, "defaultvalue"));
+    }
+
+    allOpts = CallOptions.DEFAULT;
+    for (int i = 0; i < customOptionsCount; i++) {
+      allOpts = allOpts.withOption(customOptions.get(i), "value");
+    }
+
+    shuffledCustomOptions = new ArrayList<CallOptions.Key<String>>(customOptions);
+    Collections.shuffle(shuffledCustomOptions);
```

Comment:
Micronit: Should benchmarks be deterministic or non deterministic?  You might consider seeding this if you just want a generally random list.